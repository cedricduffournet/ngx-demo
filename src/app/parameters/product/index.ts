export * from './parameter-product.component';
