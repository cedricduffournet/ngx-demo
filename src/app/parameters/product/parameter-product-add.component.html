<app-product-add-view></app-product-add-view>
