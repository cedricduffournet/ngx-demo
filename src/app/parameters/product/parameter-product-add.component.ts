import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-parameter-product-add',
  templateUrl: './parameter-product-add.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ParameterProductAddComponent implements OnInit {
  constructor() {}

  ngOnInit() {}
}
