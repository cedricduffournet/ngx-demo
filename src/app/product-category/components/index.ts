export * from './product-category-add/product-category-add.component';
export * from './product-category-update/product-category-update.component';
export * from './product-category-delete/product-category-delete.component';
export * from './product-category-form/product-category-form.component';
export * from './product-category-item/product-category-item.component';
export * from './product-category-items/product-category-items.component';
