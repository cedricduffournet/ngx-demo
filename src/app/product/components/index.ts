export * from './product-add/product-add.component';
export * from './product-update/product-update.component';
export * from './product-delete/product-delete.component';
export * from './product-form/product-form.component';
export * from './product-item/product-item.component';
export * from './product-items/product-items.component';
export * from './product-details/product-details.component';
