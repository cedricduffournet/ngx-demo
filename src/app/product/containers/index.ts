export * from './product-add-view/product-add-view.component';
export * from './product-update-modal/product-update-modal.component';
export * from './product-delete-modal/product-delete-modal.component';
export * from './product-list-view/product-list-view.component';
export * from './product-selected/product-selected.component';
export * from './product-details-view/product-details-view.component';
