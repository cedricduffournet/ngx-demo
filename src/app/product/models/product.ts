import { schema } from 'normalizr';
import { ProductCategory, productCategorySchema } from '@app/product-category/models/product-category';

export interface Product {
  id: number;
  name: string;
  categories: ProductCategory[];
}

export const productSchema = new schema.Entity('products', {
  categories: [productCategorySchema]
});
