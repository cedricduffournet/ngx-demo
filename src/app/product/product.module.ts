import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { PipesModule } from '@app/shared/pipes/pipe.module';

import { ProductCategoryModule } from '@app/product-category/product-category.module';
import { ProductEffects } from '@app/product/state/effects';
import { ProductFacade } from '@app/product/state/product.facade';
import {
  ProductAddComponent,
  ProductUpdateComponent,
  ProductDeleteComponent,
  ProductFormComponent,
  ProductItemComponent,
  ProductItemsComponent,
  ProductDetailsComponent
} from '@app/product/components';
import {
  ProductAddViewComponent,
  ProductUpdateModalComponent,
  ProductDeleteModalComponent,
  ProductListViewComponent,
  ProductSelectedComponent,
  ProductDetailsViewComponent
} from '@app/product/containers';
import { reducers } from '@app/product/state/reducers';
import { ProductGuard } from '@app/product/services/product.guard';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    PipesModule,
    ProductCategoryModule,
    EffectsModule.forFeature([ProductEffects]),
    StoreModule.forFeature('products', reducers)
  ],
  declarations: [
    ProductListViewComponent,
    ProductAddComponent,
    ProductUpdateComponent,
    ProductDeleteComponent,
    ProductFormComponent,
    ProductItemComponent,
    ProductItemsComponent,
    ProductAddViewComponent,
    ProductUpdateModalComponent,
    ProductDeleteModalComponent,
    ProductDetailsComponent,
    ProductSelectedComponent,
    ProductDetailsViewComponent
  ],
  entryComponents: [
    ProductUpdateModalComponent,
    ProductDeleteModalComponent
  ],
  exports: [ProductListViewComponent, ProductAddViewComponent, ProductDetailsViewComponent],
  providers: [ProductFacade, ProductGuard]
})
export class ProductModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
