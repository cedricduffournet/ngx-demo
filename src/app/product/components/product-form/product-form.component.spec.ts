import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ReactiveFormsModule } from '@angular/forms';
import { TranslateModule } from '@ngx-translate/core';

import { FormModule } from '@app/shared/form';
import { ValidationActionModule } from '@app/shared/validation-action';
import { ProductFormComponent } from '@app/product/components';
import { Product } from '@app/product/models/product';

describe('ProductForm', () => {
  let fixture: ComponentFixture<ProductFormComponent>;
  let component: ProductFormComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        ReactiveFormsModule,
        TranslateModule.forRoot(),
        FormModule,
        ValidationActionModule
      ],
      declarations: [ProductFormComponent]
    });

    fixture = TestBed.createComponent(ProductFormComponent);
    component = fixture.componentInstance;
  });

  it('should be created', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should not emit event if name not filled', () => {
    const product = {
      name: ''
    } as Product;

    fixture.detectChanges();
    component.productForm.setValue(product);

    spyOn(component.save, 'emit');
    component.onSubmit();

    expect(component.save.emit).not.toHaveBeenCalled();
  });


  it('should disable the form if processing', () => {
    component.processing = true;

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should emit save event with new product, if the form is valid, when submitted', () => {
    const product = {name: 'test' };
    fixture.detectChanges();

    component.productForm.setValue(product);

    spyOn(component.save, 'emit');
    component.onSubmit();

    expect(component.save.emit).toHaveBeenCalledWith(product);
  });

  it('should emit cancel event when click cancel button', () => {
    fixture.detectChanges();

    spyOn(component.cancel, 'emit');

    component.onCancel();

    expect(component.cancel.emit).toHaveBeenCalledWith('cancel');
  });

  it('should fill form with product value', () => {
    const product = {
      name: 'testName'
    } as Product;
    component.product = product;
    fixture.detectChanges();

    expect(component.productForm.value).toEqual(product);
  });
});
