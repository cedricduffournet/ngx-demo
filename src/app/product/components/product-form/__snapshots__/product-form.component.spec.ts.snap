// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm should be created 1`] = `
<app-product-form
  cancel={[Function EventEmitter]}
  categories={[Function Array]}
  formBuilder={[Function FormBuilder]}
  processing="false"
  product={[Function Object]}
  productForm={[Function FormGroup]}
  save={[Function EventEmitter]}
  selectedCategories={[Function Array]}
  submitted="false"
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <section
      class="bg-white"
    >
      <div
        class="p-3"
      >
        <app-text-input
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="name"
          ng-reflect-label="NAME"
          ng-reflect-name="name"
          ng-reflect-submitted="false"
        >
          <div
            class="form-group "
          >
            <label
              class="control-label"
            >
              NAME
            </label>
            <input
              class="form-control"
              placeholder=""
              type="text"
            />
            
            
            
            
            
            
          </div>
        </app-text-input>
        
        <app-text-input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="amount"
          ng-reflect-label="PRICE"
          ng-reflect-name="amount"
          ng-reflect-submitted="false"
          ng-reflect-type="number"
          type="number"
        >
          <div
            class="form-group "
          >
            <label
              class="control-label"
            >
              PRICE
            </label>
            <input
              class="form-control"
              placeholder=""
              type="number"
            />
            
            
            
            
            
            
          </div>
        </app-text-input>
        <app-text-area
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="description"
          ng-reflect-label="DESCRIPTION"
          ng-reflect-name="description"
          ng-reflect-submitted="false"
        >
          <div
            class="form-group "
          >
            <label
              class="control-label"
            >
              DESCRIPTION
            </label>
            <textarea
              class="form-control"
              placeholder=""
              rows="5"
            >
              
    
            </textarea>
            
            
            
          </div>
        </app-text-area>
        
        <div
          class="form-check"
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="category0"
            ng-reflect-form="[object Object]"
            type="checkbox"
            value=""
          />
          <label
            class="form-check-label"
            for="category0"
          >
             name 1 
          </label>
        </div>
        <div
          class="form-check"
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="category1"
            ng-reflect-form="[object Object]"
            type="checkbox"
            value=""
          />
          <label
            class="form-check-label"
            for="category1"
          >
             name 2 
          </label>
        </div>
      </div>
    </section>
    <app-wrapper-validation-action>
      <div
        class="d-flex justify-content-end align-items-center p-3"
      >
        <app-validation-button
          ng-reflect-confirm-disabled="false"
          ng-reflect-label-cancel="CANCEL"
          ng-reflect-label-confirm="SAVE"
        >
          <div
            class="button-validation-action text-center d-flex justify-content-end align-items-center p-3-group"
            ng-reflect-klass="button-validation-action text-"
            ng-reflect-ng-class=""
          >
            
            <button
              class="btn btn-cancel btn-lg btn-light"
              ng-reflect-klass="btn btn-cancel btn-lg"
              ng-reflect-ng-class="btn-light"
              type="button"
            >
               CANCEL 
            </button>
            
            <button
              class="btn btn-confirm btn-lg btn-primary"
              ng-reflect-klass="btn btn-confirm btn-lg"
              ng-reflect-ng-class="btn-primary"
              type="submit"
            >
              
              
              
              <span>
                SAVE
              </span>
            </button>
          </div>
        </app-validation-button>
      </div>
    </app-wrapper-validation-action>
  </form>
</app-product-form>
`;
